plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/credentials/sso/SsoAccessTokenProvider.kt

```diff
@@ -443,39 +472,109 @@ class SsoAccessTokenProvider(
+            throw e
         }
 
     private fun saveClientRegistration(registration: ClientRegistration) {
-        when (registration) {
-            is DeviceAuthorizationClientRegistration -> {
-                cache.saveClientRegistration(dagClientRegistrationCacheKey, registration)
-            }
+        val credentialType: String
+        try {
+            when (registration) {
+                is DeviceAuthorizationClientRegistration -> {
+                    credentialType = DeviceAuthorizationClientRegistration::class.java.name
+                    cache.saveClientRegistration(dagClientRegistrationCacheKey, registration)
+                }
 
-            is PKCEClientRegistration -> {
-                cache.saveClientRegistration(pkceClientRegistrationCacheKey, registration)
+                is PKCEClientRegistration -> {
+                    credentialType = PKCEClientRegistration::class.java.name
+                    cache.saveClientRegistration(pkceClientRegistrationCacheKey, registration)
+                }
             }
+        } catch (e: Exception) {
+            AwsTelemetry.createCredentials(
+                result = Result.Failed,
+                reason = "Failed to save client registration to cache",
+                reasonDesc = e.message
+            )
+            throw e
         }
+        AwsTelemetry.createCredentials(
+            result = Result.Succeeded,
+            reason = "$credentialType successfully written to cache",
+        )
     }
 
     private fun invalidateClientRegistration() {
-        cache.invalidateClientRegistration(dagClientRegistrationCacheKey)
-        cache.invalidateClientRegistration(pkceClientRegistrationCacheKey)
+        try {
+            cache.invalidateClientRegistration(dagClientRegistrationCacheKey)
+            cache.invalidateClientRegistration(pkceClientRegistrationCacheKey)
+        } catch (e: IOException) {
+            AwsTelemetry.modifyCredentials(
+                credentialModification = CredentialModification.Delete,
+                result = Result.Failed,
+                reason = "Failed to invalidate client registration",
+                reasonDesc = e.message,
+                source = "SsoAccessTokenProvider.invalidateClientRegistration"
+            )
+        }
     }
 
-    private fun saveAccessToken(token: AccessToken) {
-        when (token) {
-            is DeviceAuthorizationGrantToken -> {
-                cache.saveAccessToken(dagAccessTokenCacheKey, token)
+    private fun reauthExpiredRegistration(expiredToken: AccessToken): ClientRegistration? {
+        when (expiredToken) {
+            is DeviceAuthorizationGrantToken -> registerDAGClient()
+            is PKCEAuthorizationGrantToken -> registerPkceClient()
+        }
+        try {
+            return when (expiredToken) {
+                is DeviceAuthorizationGrantToken -> loadDagClientRegistration()
+                is PKCEAuthorizationGrantToken -> loadPkceClientRegistration()
```

Comment:
if we immediately generated a registration right before this, why do we need to try loading it from disk?